enhancements/monitoring/integration-with-topology-modes.md

```diff
@@ -0,0 +1,90 @@
+---
+title: monitoring-topology-mode
+authors:
+  - "@lilic"
+reviewers:
+  - "monitoring-team"
+  - "???"
+approvers:
+  - "monitoring-team"
+creation-date: 2021-03-01
+last-updated: 2021-03-01
+status: implementable
+see-also:
+  - "/enhancements/cluster-high-availability-mode-api.md"
+---
+
+# Monitoring Stack on Single Replica Topology Mode
+
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The monitoring stack needs to react based on the `infrastructureTopology` field newly added to `infrastructures.config.openshift.io` CRD. Plan is to detect whenever
+cluster is in single replica topology mode and reduce the monitoring stack footprint by adjusting platform monitoring
+instances to be non Highly Available.
+
+## Motivation
+
+In the 4.8 time frame, OpenShift will introduce a new capabilities field in the Infrastructure API that will inform
+operators about the deployment topology. The immediate goal is for operators to adjust the deployment of their operands
+depending on whether the infrastructure is HA (at least 3 nodes) or not (single node).
+
+### Goals
+
+- Reduce monitoring footprint by lowering number of replicas
+
+### Non-Goals
+
+- User workload stack or its Prometheus instance is not adjusted
+- Reduce memory footprint
+- Address CRC requirements (though it might help)
+- Tests from monitoring team
+
+## Proposal
+
+To adjust to the single replica mode, in cluster-monitoring-operator we detect the topology mode based on the "cluster"
+named instance of `Infrastructure`, if that mode is `SingleReplicaTopologyMode`, we only spin up one instance of
+platform Prometheus for that cluster.
+
+### Risks and Mitigations
+
+There are risks that this is not enough to reduce the actual footprint of the monitoring stack. The other thing is that
+as a result of this monitoring is no longer highly available but there is no workaround for that as its a single node
+cluster.
```

Comment:
Currently by default the metrics disappearing in case of a reboot? Or where there are 2 replicas there are synced between 2 Prometheus instances? In a multi node scenario.
If it's the later, we'll be come complete stateless in case of SNO. Unless the user specify a pvc, correct?

Reply:
In HA mode, the metrics aren't synced between Prometheus instances. The assumption is that when one Prometheus instance is down (reboot or pod restart), the other one is still up (we have anti-affinity rules ensuring that Prometheus is scheduled on different nodes). And we use Thanos querier as the main API entrypoint, Thanos pulls/aggregates data from all Prometheus instances.

Reply:
Like Simon said for Prometheus this is the case, but different for other components e.g. Alertmanager. Should we include this in the proposal or do we think its known fact?

Reply:
> Like Simon said for Prometheus this is the case, but different for other components e.g. Alertmanager. Should we include this in the proposal or do we think its known fact?

Including background information like that helps reviewers who are less familiar with some of the existing implementation details (such as me!). Ideally, this enhancement would address the impact of the change on all of the components managed by the operator.